pkgs/dkg/drand.go

```diff
@@ -159,27 +196,181 @@ func (o *LocalOwner) StartDKG() error {
+
+		OldNodes = append(OldNodes, dkg.Node{
+			Index:  dkg.Index(op.ID - 1),
+			Public: p,
+		})
+	}
+	o.Logger.Debug("Staring DKG with nodes: ")
+	for _, n := range append(OldNodes, NewNodes...) {
+		o.Logger.Debug("node: ", zap.String("nodes", n.Public.String()))
+	}
+	// New protocol
+	logger := o.Logger.With(zap.Uint64("ID", o.ID))
+	p, err := wire.NewReshareProtocolOldNodes(&wire.Config{
+		Identifier: o.Data.ReqID[:],
+		Secret:     o.Data.Secret,
+		OldNodes:   OldNodes,
+		NewNodes:   NewNodes,
+		Suite:      o.Suite,
+		T:          int(o.Data.Reshare.OldT),
+		NewT:       int(o.Data.Reshare.NewT),
+		Board:      o.Board,
+		Share:      o.SecretShare,
+		Logger:     logger,
+	})
+	if err != nil {
+		return err
+	}
+
+	go func(p *dkg.Protocol, postF func(res *dkg.OptionResult) error) {
+		res := <-p.WaitEnd()
+		postF(&res)
+	}(p, o.PostReshare)
+	close(o.StartedDKG)
 	return nil
 }
 
+func (o *LocalOwner) StartReshareDKGNewNodes() error {
+	o.Logger.Info("Starting Resharing DKG ceremony at new nodes")
+	NewNodes := make([]dkg.Node, 0)
+	for _, op := range o.Data.Reshare.NewOperators {
+		if o.Exchanges[op.ID] == nil {
+			return fmt.Errorf("no operator at Exchanges")
+		}
+		e := o.Exchanges[op.ID]
+		p := o.Suite.G1().Point()
+		if err := p.UnmarshalBinary(e.PK); err != nil {
+			return err
+		}
+
+		NewNodes = append(NewNodes, dkg.Node{
+			Index:  dkg.Index(op.ID - 1),
+			Public: p,
+		})
+	}
+	OldNodes := make([]dkg.Node, 0)
+	var commits []byte
+	var coefs []kyber.Point
+	for _, op := range o.Data.Reshare.OldOperators {
+		if o.Exchanges[op.ID] == nil {
+			return fmt.Errorf("no operator at Exchanges")
+		}
+		e := o.Exchanges[op.ID]
+		if e.Commits == nil {
+			return fmt.Errorf("no commits at Exchanges")
+		}
+		o.Logger.Debug("Commits at exchange", zap.Uint64("ID", op.ID), zap.Binary("commits", e.Commits))
+		commits = e.Commits
+		p := o.Suite.G1().Point()
+		if err := p.UnmarshalBinary(e.PK); err != nil {
+			return err
+		}
+
+		OldNodes = append(OldNodes, dkg.Node{
+			Index:  dkg.Index(op.ID - 1),
+			Public: p,
+		})
+	}
+	o.Logger.Debug("Staring resharing DKG with nodes: ")
+	for _, n := range append(OldNodes, NewNodes...) {
+		o.Logger.Debug("node: ", zap.String("nodes", n.Public.String()))
+	}
+	coefsBytes := utils.SplitBytes(commits, 48)
+	for _, c := range coefsBytes {
+		p := o.Suite.G1().Point()
+		err := p.UnmarshalBinary(c)
+		if err != nil {
+			return err
+		}
+		coefs = append(coefs, p)
+	}
+
+	// New protocol
+	logger := o.Logger.With(zap.Uint64("ID", o.ID))
+	p, err := wire.NewReshareProtocolNewNodes(&wire.Config{
+		Identifier:   o.Data.ReqID[:],
+		Secret:       o.Data.Secret,
+		OldNodes:     OldNodes,
+		NewNodes:     NewNodes,
+		Suite:        o.Suite,
+		T:            int(o.Data.Reshare.OldT),
+		NewT:         int(o.Data.Reshare.NewT),
+		Board:        o.Board,
+		PublicCoeffs: coefs,
+		Logger:       logger,
+	})
+	if err != nil {
+		return err
+	}
+	for _, b := range o.Deals {
+		o.Board.DealC <- *b
```

Comment:
I don't get how this works honestly as from what I se in the code of `board.go` this channel is unbuffered. I guess starting the protocol reads these from `p`, but it means it might be smarter to run the go routine underneath before pushing into the channel.

Reply:
well I guess `NewReshareProtocolNewNodes` listen to this so no need to run the goroutine before.